rclpy/test/test_node.py

```diff
@@ -1212,6 +1238,184 @@ def test_set_descriptor_read_only(self):
                 )
             )
 
+    def test_floating_point_range_descriptor(self):
+        # OK cases; non-floats are not affected by the range.
+        fp_range = FloatingPointRange(from_value=0.0, to_value=10.0, step=0.5)
+        parameters = [
+            ('from_value', 0.0, ParameterDescriptor(floating_point_range=[fp_range])),
+            ('to_value', 10.0, ParameterDescriptor(floating_point_range=[fp_range])),
+            ('in_range', 4.5, ParameterDescriptor(floating_point_range=[fp_range])),
+            ('str_value', 'I am no float', ParameterDescriptor(floating_point_range=[fp_range])),
+            ('int_value', 123, ParameterDescriptor(floating_point_range=[fp_range]))
+        ]
+
+        result = self.node.declare_parameters('', parameters)
+
+        self.assertIsInstance(result, list)
+        self.assertIsInstance(result[0], Parameter)
+        self.assertIsInstance(result[1], Parameter)
+        self.assertIsInstance(result[2], Parameter)
+        self.assertIsInstance(result[3], Parameter)
+        self.assertIsInstance(result[4], Parameter)
+        self.assertAlmostEqual(result[0].value, 0.0)
+        self.assertAlmostEqual(result[1].value, 10.0)
+        self.assertAlmostEqual(result[2].value, 4.5)
+        self.assertEqual(result[3].value, 'I am no float')
+        self.assertEqual(result[4].value, 123)
+        self.assertEqual(self.node.get_parameter('from_value').value, 0.0)
+        self.assertEqual(self.node.get_parameter('to_value').value, 10.0)
+        self.assertEqual(self.node.get_parameter('in_range').value, 4.5)
+        self.assertEqual(self.node.get_parameter('str_value').value, 'I am no float')
+        self.assertAlmostEqual(self.node.get_parameter('int_value').value, 123)
+
+        # Try to set a parameter out of range.
+        result = self.node.set_parameters([Parameter('in_range', value=12.0)])
+        self.assertIsInstance(result, list)
+        self.assertIsInstance(result[0], SetParametersResult)
+        self.assertFalse(result[0].successful)
+        self.assertEqual(self.node.get_parameter('in_range').value, 4.5)
+
+        # Try to set a parameter out of range (bad step).
+        result = self.node.set_parameters([Parameter('in_range', value=4.25)])
+        self.assertIsInstance(result, list)
+        self.assertIsInstance(result[0], SetParametersResult)
+        self.assertFalse(result[0].successful)
+        self.assertEqual(self.node.get_parameter('in_range').value, 4.5)
+
+        # Change in_range parameter to int; ranges will not apply.
+        result = self.node.set_parameters([Parameter('in_range', value=12)])
```

Comment:
That's subtle. I hope users don't forget (or accidentally add) a decimal point in their code, otherwise they may experience unexpected behavior. I guess this is a case for type constraints.

Reply:
Indeed.
Note that in this case I'm making the `Parameter` guess the type; using the old API where the type had to be explicitly given to the parameter you wouldn't have this problem.

In any case, I also think the solution is eventually enforcing the types (now for Dashing, or later for a patch release).